File: Hard/2026-03-15-1622-Fancy-Sequence/solution.py
class Fancy:
    def __init__(self):
        self.sequence = []
        self.add = 0
        self.mul = 1
        self.MOD = 10**9 + 7

    def append(self, val: int) -> None:
        val = (val - self.add) * pow(self.mul, self.MOD - 2, self.MOD) % self.MOD
        self.sequence.append(val)

    def addAll(self, inc: int) -> None:
        self.add = (self.add + inc) % self.MOD

    def multAll(self, m: int) -> None:
        self.add = self.add * m % self.MOD
        self.mul = self.mul * m % self.MOD

    def getIndex(self, idx: int) -> int:
        if idx >= len(self.sequence):
            return -1
        return (self.sequence[idx] * self.mul + self.add) % self.MOD

File: Hard/2026-03-15-1622-Fancy-Sequence/solution.js
class Fancy {
    constructor() {
        this.sequence = [];
        this.add = 0;
        this.mul = 1;
    }

    append(val) {
        val = (val - this.add + 10**9 + 7) % (10**9 + 7);
        this.sequence.push((val * this.mul) % (10**9 + 7));
    }

    addAll(inc) {
        this.add = (this.add + inc) % (10**9 + 7);
    }

    multAll(m) {
        this.add = (this.add * m) % (10**9 + 7);
        this.mul = (this.mul * m) % (10**9 + 7);
    }

    getIndex(idx) {
        if (idx >= this.sequence.length) return -1;
        return (this.sequence[idx] + this.add) % (10**9 + 7);
    }
}